File: Tests/TarotMCPTests/TestSupport/CountedSet.swift
struct CountedSet<Element: Hashable> {
  private var storage: [Element: Int] = [:]

  mutating func insert(_ element: Element) {
    storage[element, default: 0] += 1
  }

  mutating func remove(_ element: Element) {
    guard let currentCount = storage[element], currentCount > 0 else { return }
    if currentCount == 1 {
      storage.removeValue(forKey: element)
    } else {
      storage[element] = currentCount - 1
    }
  }

  func count(of element: Element) -> Int {
    return storage[element] ?? 0
  }
}

File: src/__tests__/test-support/counted-set.ts
/**
 * TypeScript equivalent of Swift's CountedSet for tracking element frequencies
 */
export class CountedSet<T> {
  private storage = new Map<T, number>();

  insert(element: T): void {
    const currentCount = this.storage.get(element) ?? 0;
    this.storage.set(element, currentCount + 1);
  }

  remove(element: T): void {
    const currentCount = this.storage.get(element);
    if (currentCount === undefined || currentCount <= 0) {
      return;
    }
    if (currentCount === 1) {
      this.storage.delete(element);
    } else {
      this.storage.set(element, currentCount - 1);
    }
  }

  count(element: T): number {
    return this.storage.get(element) ?? 0;
  }

  get size(): number {
    return this.storage.size;
  }

  clear(): void {
    this.storage.clear();
  }
} 